README.md
# jQuery Palette

Palette is a simple, clean, color picker plugin for [jQuery][jQuery] v1.7+.


## Getting Started

First, make sure the required JavaScript files are included in the `<head>` tag.

~~~ { html }
<script src="js/parse-color.min.js" type="text/javascript"></script>
<script src="js/jquery.min.js" type="text/javascript"></script>
<script src="js/jquery.palette.min.js" type="text/javascript"></script>
~~~

Next, include Palette's stylesheet _(you may need to update the image paths depending on how your site is structured)_.

~~~ { html }
<link rel="stylesheet" href="css/palette.css" type="text/css">
~~~

Lastly, attach Palette to which ever element(s) you want.

~~~ { javascript }
$('#palette').palette({'color': 'CC4747'});
~~~


## Settings

<table>
    <thead>
        <tr>
            <th>Property</th>
            <th>Type</th>
            <th>Default</th>
            <th>Description</th>
        </tr>
    </thead>
    
    <tbody>
        <tr>
            <td><code>color</code></td>
            <td><em>string|object</em></td>
            <td><em>'000000'</em></td>
            <td>
                <p>The default color of the picker.</p>
                <p><em>Note, can be in any format <a href="https://github.com/Anomareh/parse-color.js">parse-color.js</a> recognizes.</em></p>
            </td>
        </tr>
        
        <tr>
            <td><code>onChange</code></td>
            <td><em>function</em></td>
            <td>&mdash;</td>
            <td>
                <p>A callback function that's fired when the color of the picker changes.</p>
                
                <p><em>Note, this callback is <strong>not</strong> fired for live changes (e.g. while <strong>dragging</strong> the selector cursor as well as the hue and field sliders). For that use <code>onLive</code>.</em></p>
            </td>
        </tr>
        
        <tr>
            <td><code>onHide</code></td>
            <td><em>function</em></td>
            <td>&mdash;</td>
            <td>A callback function that's fired when the picker is hidden.</td>
        </tr>
        
        <tr>
            <td><code>onLive</code></td>
            <td><em>function</em></td>
            <td>&mdash;</td>
            <td>
                <p>A callback function that's fired <strong>immediately</strong> when the color of the picker changes.</p>
                
                <p><em>Note, this callback is <strong>only</strong> fired for live changes (e.g. while <strong>dragging</strong> the selector cursor as well as the hue and field sliders). <code>onChange</code> will fire on the following <code>mouseup</code> after dragging.</em></p>
            </td>
        </tr>
        
        <tr>
            <td><code>onShow</code></td>
            <td><em>function</em></td>
            <td>&mdash;</td>
            <td>A callback function that's fired when the picker is shown.</td>
        </tr>
        
        <tr>
            <td><code>onSubmit</code></td>
            <td><em>function</em></td>
            <td>&mdash;</td>
            <td>A callback function that's fired when the <strong>current</strong> color of the picker is changed <em>(e.g. when the submit button is clicked)</em>.</td>
        </tr>
    </tbody>
</table>


## Methods

<table>
    <thead>
        <tr>
            <th>Method</th>
            <th>Description</th>
        </tr>
    </thead>
    
    <tbody>
        <tr>
            <td><code>$.palette.color($palette, color, 'from')</code></td>
            <td>
                <p>Gets or sets the color of the supplied picker.</p>
                
                <p><code>color</code> can be in any format <a href="https://github.com/Anomareh/parse-color.js">parse-color.js</a> recognizes. If it isn't supplied, the color of the supplied picker is returned.</p>
                
                <p><code>from</code> is a string telling Palette where the color of the picker is being changed from and can be anything. For the most part, it is used internally to tell Palette not to overwrite the field the color is being changed from. If it isn't supplied, all fields are updated.</p>
            </td>
        </tr>
        
        <tr>
            <td><code>$.palette.hide()</code></td>
            <td>Hides the picker(s).</td>
        </tr>
        
        <tr>
            <td><code>$.palette.show($palette)</code></td>
            <td>Shows the supplied picker.</td>
        </tr>
    </tbody>
</table>


## Data

~~~ { javascript }
$('#palette').palette();
~~~

Given the above usage of Palette, the following [data][jQuery-data] is available.

### Parent

~~~ { javascript }
var parent_data = $('#palette').data('palette');
~~~

<table>
    <thead>
        <tr>
            <th>Property</th>
            <th>Description</th>
        </tr>
    </thead>
    
    <tbody>
        <tr>
            <td><code>palette</code></td>
            <td>The jQuery object of the picker.</td>
        </tr>
    </tbody>
</table>

### Picker

~~~ { javascript }
var
    $palette = parent_data.palette,
    
    picker_data = $palette.data('palette');
~~~

<table>
    <thead>
        <tr>
            <th>Property</th>
            <th>Description</th>
        </tr>
    </thead>
    
    <tbody>
        <tr>
            <td><code>color</code></td>
            <td>
                <p>The <a href="https://github.com/Anomareh/parse-color.js">parse-color.js</a> object of the color of the picker.</p>
                
                <p><em>Note, this can differ from <code>current</code>.</em></p>
            </td>
        </tr>
        
        <tr>
            <td><code>current</code></td>
            <td>The <a href="https://github.com/Anomareh/parse-color.js">parse-color.js</a> object of the <strong>current</strong> color of the picker.</td>
        </tr>
        
        <tr>
            <td><code>parent</code></td>
            <td>The jQuery object of the element the picker is attached to.</td>
        </tr>
    </tbody>
</table>

_Also note, all of the picker's options are available via it's data._


## Dependencies

+ [jQuery v1.7+][jQuery]
+ [parse-color.js][parse-color.js]


[jQuery]: http://jquery.com/
[jQuery-data]: http://api.jquery.com/data/
[parse-color.js]: https://github.com/Anomareh/parse-color.js


CHANGELOG.md
# Changelog

### v0.1 (January 22, 2012)

+ Initial release.


src/parse-color.js
function parse_color(color) {
    var
        hex = '000000',
        hsl = {'h': 0, 's': 0, 'l': 0},
        hsv = {'h': 0, 's': 0, 'v': 0},
        luma = 0,
        rgb = {'r': 0, 'g': 0, 'b': 0},
        
        definitions = [
            {
                're': /^([0-9a-f]{1})([0-9a-f]{1})([0-9a-f]{1})$/,
                'process': function(bits) {
                    return {
                        'r': parseInt(bits[1] + bits[1], 16),
                        'g': parseInt(bits[2] + bits[2], 16),
                        'b': parseInt(bits[3] + bits[3], 16)
                    };
                }
            },
            {
                're': /^([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})$/,
                'process': function(bits) {
                    return {
                        'r': parseInt(bits[1], 16),
                        'g': parseInt(bits[2], 16),
                        'b': parseInt(bits[3], 16)
                    };
                }
            },
            {
                're': /^rgb\((\d{1,3}),\s*(\d{1,3}),\s*(\d{1,3})\)$/,
                'process': function(bits) {
                    return {
                        'r': parseInt(bits[1], 10),
                        'g': parseInt(bits[2], 10),
                        'b': parseInt(bits[3], 10)
                    };
                }
            },
            {
                're': /^rgba\((\d{1,3}),\s*(\d{1,3}),\s*(\d{1,3}),\s*(?:[0-9]+(?:\.[0-9]+)?|\.[0-9]+)\)$/,
                'process': function(bits) {
                    return {
                        'r': parseInt(bits[1], 10),
                        'g': parseInt(bits[2], 10),
                        'b': parseInt(bits[3], 10)
                    };
                }
            }
        ],
        field_max = {
            'h': 360,
            's': 100,
            'l': 100,
            'v': 100,
            'r': 255,
            'g': 255,
            'b': 255
        };
    
    var check_model = function() {
        for (field in color) {
            color[field] = (color[field] < 0 || isNaN(color[field])) ? 0 : ((color[field] > field_max[field]) ? field_max[field] : color[field]);
        }
        
        return color;
    };
    
    var get_luma = function() {
        luma = (rgb.r * 0.21 + rgb.g * 0.72 + rgb.b * 0.07) / 255;
    };
    
    var hsl_to_hsv = function() {
        var
            s = hsl.s / 100,
            l = hsl.l / 100,
            
            c = s * (1 - Math.abs(2 * l - 1));
        
        hsv.h = hsl.h;
        hsv.s = 0;
        hsv.v = (2 * l + c) / 2;
        
        if (c != 0) {
            hsv.s = c / hsv.v;
        }
        
        hsv.s *= 100;
        hsv.v *= 100;
    };
    
    var hsv_to_hsl = function() {
        var
            s = hsv.s / 100,
            v = hsv.v / 100,
            
            max = v,
            min = v * (-s + 1),
            
            c = max - min;
        
        hsl.h = hsv.h;
        hsl.s = 0;
        hsl.l = (max + min) / 2;
        
        if (c != 0) {
            hsl.s = c / (1 - Math.abs(2 * hsl.l - 1));
        }
        
        hsl.s *= 100;
        hsl.l *= 100;
    };
    
    var hsv_to_rgb = function() {
        var
            h = hsv.h,
            s = hsv.s / 100,
            v = hsv.v / 100;
        
        rgb.b = rgb.g = rgb.r = v;
        
        if (s != 0) {
            var
                sector_pos = h / 60,
                
                sector_num = Math.floor(sector_pos),
                
                sector_frac = sector_pos - sector_num,
                
                p = v * (1 - s),
                q = v * (1 - (s * sector_frac)),
                t = v * (1 - (s * (1 - sector_frac)));
            
            switch (sector_num) {
                case 0:
                case 6:
                    rgb.g = t;
                    rgb.b = p;
                    
                    break;
                case 1:
                    rgb.r = q;
                    rgb.b = p;
                    
                    break;
                case 2:
                    rgb.r = p;
                    rgb.b = t;
                    
                    break;
                case 3:
                    rgb.r = p;
                    rgb.g = q;
                    
                    break;
                case 4:
                    rgb.r = t;
                    rgb.g = p;
                    
                    break;
                case 5:
                    rgb.g = p;
                    rgb.b = q;
                    
                    break;
            }
        }
        
        rgb.r = Math.round(rgb.r * 255);
        rgb.g = Math.round(rgb.g * 255);
        rgb.b = Math.round(rgb.b * 255);
    };
    
    var rgb_to_hex = function() {
        var
            r = rgb.r.toString(16),
            g = rgb.g.toString(16),
            b = rgb.b.toString(16);
        
        r = (r.length == 1) ? '0' + r : r;
        g = (g.length == 1) ? '0' + g : g;
        b = (b.length == 1) ? '0' + b : b;
        
        hex = (r + g + b).toUpperCase();
    };
    
    var rgb_to_hslv = function() {
        var
            r = rgb.r / 255,
            g = rgb.g / 255,
            b = rgb.b / 255,
            
            max = Math.max(r, g, b),
            min = Math.min(r, g, b),
            
            c = max - min;
        
        hsl.s = hsl.h = max;
        hsl.l = (max + min) / 2;
        
        hsv.v = hsv.s = hsv.h = max;
        
        switch (max) {
            case r:
                if (g >= b) {
                    if (c == 0) {
                        hsl.h = 0;
                    } else {
                        hsl.h = 60 * (g - b) / c;
                    }
                } else if (g < b) {
                    hsl.h = 60 * (g - b) / c + 360;
                }
                
                break;
            case g:
                hsl.h = 60 * (b - r) / c + 120;
                
                break;
            case b:
                hsl.h = 60 * (r - g) / c + 240;
                
                break;
        }
        
        hsv.h = hsl.h;
        
        if (c == 0) {
            hsv.s = hsl.s = 0;
        } else {
            hsl.s = c / (1 - Math.abs(2 * hsl.l - 1));
            hsv.s = c / hsv.v;
        }
        
        hsl.s *= 100;
        hsl.l *= 100;
        hsv.s *= 100;
        hsv.v *= 100;
    };
    
    var str_to_rgb = function() {
        color = color.toLowerCase().replace(/[\s#]/g, '');
        
        for (var i = 0; i < definitions.length; i++) {
            var
                processor = definitions[i].process,
                re = definitions[i].re,
                
                bits = re.exec(color);
            
            if (bits) {
                color = processor(bits);
                
                break;
            }
        }
        
        if (typeof color == 'object') {
            rgb = check_model();
        }
    };
    
    switch (typeof color) {
        case 'number':
            var h = parseInt(color, 10);
            
            hsv.h = (isNaN(h)) ? 0 : Math.max(Math.min(h, 360), 0);
            hsv.s = Math.round(Math.random() * 100);
            hsv.v = Math.round(Math.random() * 100);
            
            hsv_to_hsl();
            hsv_to_rgb();
            rgb_to_hex();
            get_luma();
            
            break;
        case 'object':
            if ('r' in color && 'g' in color && 'b' in color) {
                rgb = check_model();
                
                rgb_to_hslv();
            } else if ('h' in color && 's' in color) {
                if ('l' in color) {
                    hsl = check_model();
                    
                    hsl_to_hsv();
                } else {
                    hsv = check_model();
                    
                    hsv_to_hsl();
                }
                
                hsv_to_rgb();
            } else {
                break;
            }
            
            rgb_to_hex();
            get_luma();
            
            break;
        case 'string':
            str_to_rgb();
            rgb_to_hslv();
            rgb_to_hex();
            get_luma();
            
            break;
        default:
            rgb.r = Math.round(Math.random() * 255);
            rgb.g = Math.round(Math.random() * 255);
            rgb.b = Math.round(Math.random() * 255);
            
            rgb_to_hex();
            rgb_to_hslv();
            get_luma();
            
            break;
    }
    
    return {
        'hex': hex,
        'hsl': hsl,
        'hsv': hsv,
        'luma': luma,
        'rgb': rgb
    };
}


src/jquery.palette.js
(function($, window, document) {
    var
        $document = $(document),
        $window = $(window),
        
        defaults = {
            'color': '000000',
            'onChange': function() {},
            'onHide': function() {},
            'onLive': function() {},
            'onShow': function() {},
            'onSubmit': function() {}
        },
        fields = {
            4: {
                'field': 'r',
                'max': 255,
                'model': 'rgb'
            },
            5: {
                'field': 'g',
                'max': 255,
                'model': 'rgb'
            },
            6: {
                'field': 'b',
                'max': 255,
                'model': 'rgb'
            },
            7: {
                'field': 'h',
                'max': 360,
                'model': 'hsv'
            },
            8: {
                'field': 's',
                'max': 100,
                'model': 'hsv'
            },
            9: {
                'field': 'v',
                'max': 100,
                'model': 'hsv'
            }
        },
        html = '<div class="palette"><div class="palette_selector"><div><div>&nbsp;</div></div></div><div class="palette_hue"><div class="palette_hue_slider">&nbsp;</div><div>&nbsp;</div></div><div class="palette_new_color"><div>&nbsp;</div></div><div class="palette_current_color"><div>&nbsp;</div></div><div class="palette_field palette_r"><input type="text" maxlength="3"><span>&nbsp;</span></div><div class="palette_field palette_g"><input type="text" maxlength="3"><span>&nbsp;</span></div><div class="palette_field palette_b"><input type="text" maxlength="3"><span>&nbsp;</span></div><div class="palette_field palette_h"><input type="text" maxlength="3"><span>&nbsp;</span></div><div class="palette_field palette_s"><input type="text" maxlength="3"><span>&nbsp;</span></div><div class="palette_field palette_v"><input type="text" maxlength="3"><span>&nbsp;</span></div><div class="palette_hex"><input type="text" maxlength="6"></div><div class="palette_submit">&nbsp;</div></div>';
    
    var add_bindings = function($palette) {
        var data = $palette.data('palette');
        
        // selector
        $palette.find('.palette_selector > div').mousedown(function(e) {
            if (!(e.which > 1 || e.altKey || e.ctrlKey || e.metaKey || e.shiftKey)) {
                e.preventDefault();
                
                var $selector = $(this);
                
                change_selector($selector, e.pageX, e.pageY);
                
                $document.on('mousemove.palette', function(e) {
                    change_selector($selector, e.pageX, e.pageY);
                }).one('mouseup.palette', function() {
                    $document.off('mousemove.palette');
                    
                    data.onChange($palette);
                });
            }
        });
        
        // hue
        $palette.children('.palette_hue').mousedown(function(e) {
            if (!(e.which > 1 || e.altKey || e.ctrlKey || e.metaKey || e.shiftKey)) {
                e.preventDefault();
                
                var $hue = $(this);
                
                change_hue($hue, e.pageY);
                
                $document.on('mousemove.palette', function(e) {
                    change_hue($hue, e.pageY);
                }).one('mouseup.palette', function() {
                    $document.off('mousemove.palette');
                    
                    data.onChange($palette);
                });
            }
        });
        
        // current color
        $palette.children('.palette_current_color').click(function(e) {
            if (!(e.which > 1 || e.altKey || e.ctrlKey || e.metaKey || e.shiftKey)) {
                $.palette.color($palette, data.current.hsv, 'current');
            }
        });
        
        // fields
        $palette.find('.palette_field input').keyup(function(e) {
            var
                $input = $(this),
                
                val = parseInt($input.val(), 10);
            
            if (e.which == 13 || e.which == 27) {
                $input.blur();
            } else {
                var 
                    info = fields[$input.parent().index()],
                    
                    color = $.extend({}, data.color[info.model]);
                
                if (isNaN(val)) {
                    val = 0;
                } else if (val > info.max) {
                    $input.val(info.max.toString());
                    val = info.max;
                }
                
                color[info.field] = val;
                
                $.palette.color($palette, color, info.field);
            }
        }).keydown(function(e) {
            if (e.which == 38 || e.which == 40) {
                var
                    $input = $(this),
                    
                    info = fields[$input.parent().index()],
                    old = val = parseInt($input.val(), 10),
                    
                    color = $.extend({}, data.color[info.model]);
                
                if (isNaN(val)) {
                    old = val = 0;
                }
                
                if (e.which == 40 && val > 0) {
                    val--;
                    $input.val(val.toString());
                } else if (e.which == 38 && val < info.max) {
                    val++;
                    $input.val(val.toString());
                }
                
                if (val != old) {
                    color[info.field] = val;
                
                    $.palette.color($palette, color, info.field);
                }
            }
        }).blur(function() {
            var
                $input = $(this),
                
                val = parseInt($input.val(), 10);
            
            if (isNaN(val)) {
                $input.val('0');
            } else if (val.toString() != $input.val()) {
                $input.val(val.toString());
            }
        });
        
        // field sliders
        $palette.find('.palette_field span').mousedown(function(e) {
            if (!(e.which > 1 || e.altKey || e.ctrlKey || e.metaKey || e.shiftKey)) {
                e.preventDefault();
                
                var
                    $slider = $(this),
                    
                    $field = $slider.parent(),
                    
                    $input = $field.children('input'),
                    
                    info = fields[$field.index()],
                    old = parseInt($input.val(), 10),
                    y = e.pageY;
                
                $document.on('mousemove.palette', function(e) {
                    var
                        color = $.extend({}, data.color[info.model]),
                        val = Math.min(Math.max(old + y - e.pageY, 0), info.max);
                    
                    $input.val(val);
                    
                    color[info.field] = val;
                    
                    $.palette.color($palette, color, 'slider');
                }).one('mouseup.palette', function() {
                    $document.off('mousemove.palette');
                    
                    data.onChange($palette);
                });
            }
        });
        
        // hex
        $palette.find('.palette_hex input').keyup(function(e) {
            var $input = $(this);
            
            if (e.which == 13 || e.which == 27) {
                $input.blur();
            } else if ($input.val().length == 3 || $input.val().length == 6) {
                $.palette.color($palette, $input.val(), 'hex');
            }
        }).blur(function() {
            $.palette.color($palette, data.color.hsv, 'blur');
        });
        
        // submit
        $palette.children('.palette_submit').click(function(e) {
            if (!(e.which > 1 || e.altKey || e.ctrlKey || e.metaKey || e.shiftKey)) {
                $palette.find('.palette_current_color div').css('background-color', '#' + $.palette.color($palette));
                
                data.current = $.extend({}, data.color);
                
                data.onSubmit($palette);
            }
        });
    };
    
    var change_hue = function($hue, y) {
        var
            $palette = $hue.closest('.palette'),
            $hue = $hue.children('div:last'),
            
            offset = $hue.offset(),
            
            top = (y < offset.top) ? 0 : (y > (offset.top + 150)) ? 150 : y - offset.top;
        
        $hue.prev().css('top', top.toString() + 'px');
        
        $.palette.color($palette, $.extend({}, $palette.data('palette').color.hsv, {
            'h': 360 * (150 - Math.max(0, Math.min(150, top))) / 150
        }), 'hue');
    };
    
    var change_selector = function($selector, x, y) {
        var
            $palette = $selector.closest('.palette'),
            
            offset = $selector.offset(),
            
            top = (y < offset.top) ? 0 : (y > (offset.top + 150)) ? 150 : y - offset.top,
            left = (x < offset.left) ? 0 : (x > (offset.left + 150)) ? 150 : x - offset.left;
        
        $selector.children('div').css({
            'top': top.toString() + 'px',
            'left': left.toString() + 'px'
        });
        
        $.palette.color($palette, $.extend({}, $palette.data('palette').color.hsv, {
            's': 100 * (Math.max(0, Math.min(150, left))) / 150,
            'v': 100 * (150 - Math.max(0, Math.min(150, top))) / 150
        }), 'selector');
    };
    
    var compare_hsv = function(one, two) {
        for (field in one) {
            if (one[field] != two[field]) {
                return false;
            }
        }
        
        return true;
    };
    
    var position = function($palette) {
        var
            $parent = $palette.data('palette').parent,
            
            height = $parent.outerHeight(),
            offset = $parent.offset(),
            width = $parent.outerWidth();
        
        if (($window.height() - (offset.top - $window.scrollTop() + height)) > 182) {
            $palette.css('top', (offset.top + height).toString() + 'px');
        } else {
            $palette.css('top', (offset.top - 182).toString() + 'px');
        }
        
        if (($window.width() - offset.left - $window.scrollLeft()) > 367) {
            $palette.css('left', (offset.left).toString() + 'px');
        } else {
            $palette.css('left', (offset.left - width - 367).toString() + 'px');
        }
    };
    
    $.palette = {
        'color': function($palette, color, from) {
            if (typeof color == 'undefined') {
                return $palette.find('.palette_hex input').val();
            }
            
            var
                color = parse_color(color),
                data = $palette.data('palette'),
                
                old = $.extend({}, data.color),
                
                hex = color.hex,
                hsv = color.hsv,
                rgb = color.rgb,
                
                hue = parse_color({
                    'h': hsv.h,
                    's': 100,
                    'v': 100
                });
            
            data.color = color;
            
            if (from != 'selector') {
                $palette.find('.palette_selector > div').css('background-color', '#' + hue.hex);
                $palette.find('.palette_selector div div').css({
                    'top': (150 * (100 - hsv.v) / 100).toString() + 'px',
                    'left': (150 * hsv.s / 100).toString() + 'px'
                });
            }
            
            if (from != 'hue') {
                $palette.find('.palette_hue_slider').css('top', (150 - 150 * hsv.h / 360).toString() + 'px')
            }
            
            $palette.find('.palette_new_color div').css('background-color', '#' + hex);
            
            if (typeof from == 'undefined') {
                $palette.find('.palette_current_color div').css('background-color', '#' + hex);
                
                data.current = color;
                
                data.onSubmit($palette);
            }
            
            for (field in rgb) {
                if (from != field) {
                    $palette.find('.palette_' + field + ' input').val(rgb[field]);
                }
            }
            
            for (field in hsv) {
                if (from != field) {
                    $palette.find('.palette_' + field + ' input').val(Math.round(hsv[field]));
                }
            }
            
            if (from != 'hex') {
                $palette.find('.palette_hex input').val(hex);
            }
            
            if (!old.hsv || !compare_hsv(hsv, old.hsv)) {
                if (from == 'selector' || from == 'hue' || from == 'slider') {
                    data.onLive($palette);
                } else {
                    data.onChange($palette);
                }
            }
        },
        'hide': function() {
            $('.palette:visible').each(function() {
                var $palette = $(this);
                
                $palette.find('input:focus').blur();
                $palette.fadeOut(500, function() {
                    $palette.data('palette').onHide($palette);
                });
            });
        },
        'show': function($palette) {
            position($palette, $palette.data('palette').parent);
            
            $palette.fadeIn(500, function() {
                $palette.data('palette').onShow($palette);
            });
        }
    };
    
    $document.on('keyup.palette', function(e) {
        if (e.which == 27) {
            $.palette.hide();
        }
    });
    
    $.fn.palette = function(options) {
        options = $.extend({}, defaults, options || {});
        
        var
            $this = this,
            $body = $(document.body);
        
        return $this.each(function() {
            var $parent = $(this);
            
            if (!$parent.data('palette')) {
                var $palette = $(html);
                
                $body.append($palette);
                
                $palette.data('palette', $.extend({}, options, {'parent': $parent}));
                $parent.data('palette', {'palette': $palette});
                
                add_bindings($palette);
                
                $.palette.color($palette, options.color);
                
                position($palette);
                
                $parent.on('click.palette', function(e) {
                    if (!(e.which > 1 || e.altKey || e.ctrlKey || e.metaKey || e.shiftKey) && e.target == this) {
                        $.palette.hide()
                        
                        if (!$palette.is(':visible')) {
                            $.palette.show($palette);
                        }
                    }
                });
            }
        });
    };
})(jQuery, window, document);
